File: Algorithms/Medium/TimeBasedKeyValueStore.py
class TimeMap:
    def __init__(self):
        self.hash_table = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.hash_table.setdefault(key, []).append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        val = self.hash_table.get(key, None)
        if val is None:
            return ""

        l, r = 0, len(val) - 1
        while l < r:
            m = (l + r + 1) // 2
            if timestamp >= val[m][0]:
                l = m
            else:
                r = m - 1

        if val[l][0] <= timestamp:
            return val[l][1]

        return ""

File: Algorithms/Medium/TimeBasedKeyValueStore.go
package medium

type tuple struct {
	value     string
	timestamp int
}

type TimeMap struct {
	hashTable map[string][]tuple
}

func Constructor() TimeMap {
	return TimeMap{
		hashTable: make(map[string][]tuple),
	}
}

func (this *TimeMap) Set(key string, value string, timestamp int) {
	this.hashTable[key] = append(this.hashTable[key], tuple{value, timestamp})
}

func (this *TimeMap) Get(key string, timestamp int) string {
	if _, ok := this.hashTable[key]; !ok {
		return ""
	}

	l, r := 0, len(this.hashTable[key])-1
	for l < r {
		m := (l + r + 1) / 2
		if this.hashTable[key][m].timestamp <= timestamp {
			l = m
		} else {
			r = m - 1
		}
	}

	if this.hashTable[key][l].timestamp > timestamp {
		return ""
	}

	return this.hashTable[key][l].value
}
